ocaml/xapi/repository.ml

```diff
@@ -188,58 +231,136 @@ let sync ~__context ~self ~token ~token_id =
+          match yum_conf with
+          | Some (token_path, yum_plugin) ->
+              (* Configure proxy and token *)
+              let token_param =
+                match token_path with
+                | "" ->
+                    ""
+                | p ->
+                    Printf.sprintf "--setopt=%s.%s=%s" repo_name yum_plugin
+                      (Uri.make ~scheme:"file" ~path:p () |> Uri.to_string)
+              in
+              let proxy_url_param, proxy_username_param, proxy_password_param =
+                get_proxy_params ~__context repo_name
+              in
+              let Pkg_mgr.{cmd; params} =
+                [
+                  "--save"
+                ; proxy_url_param
+                ; proxy_username_param
+                ; proxy_password_param
+                ; token_param
+                ]
+                |> fun config -> Pkgs.config_repo ~repo_name ~config
+              in
+              ignore
+                (Helpers.call_script ~log_output:Helpers.On_failure cmd params)
           | None ->
-              ""
+              ()
         in
-        let proxy_url_param, proxy_username_param, proxy_password_param =
-          get_proxy_params ~__context repo_name
+
+        let make_cache () =
+          (* Import YUM repository GPG key to check metadata in reposync *)
+          let Pkg_mgr.{cmd; params} = Pkgs.make_cache ~repo_name in
+          ignore (Helpers.call_script cmd params)
         in
-        let Pkg_mgr.{cmd; params} =
-          [
-            "--save"
-          ; proxy_url_param
-          ; proxy_username_param
-          ; proxy_password_param
-          ; token_param
-          ]
-          |> fun config -> Pkgs.config_repo ~repo_name ~config
+
+        (* Sync with remote repository *)
+        let sync_repo () =
+          let Pkg_mgr.{cmd; params} = Pkgs.sync_repo ~repo_name in
+          Unixext.mkdir_rec !Xapi_globs.local_pool_repo_dir 0o700 ;
+          clean_yum_cache repo_name ;
+          ignore (Helpers.call_script cmd params)
+        in
+
+        let server_auth conf f =
+          match conf with
+          | DefaultAuth | NoAuth ->
+              f ()
+          | ServerAuthConf {cert; remote_addr; remote_port} ->
+              let ( let@ ) f x = f x in
+              let@ temp_file =
+                Helpers.with_temp_file_of_content "external-host-cert-" ".pem"
+                  cert
+              in
+              Stunnel.with_client_proxy
+                ~verify_cert:(Stunnel_client.external_host temp_file)
+                ~remote_host:remote_addr ~remote_port ~local_host ~local_port
+              @@ fun () -> f ()
         in
-        ignore (Helpers.call_script ~log_output:Helpers.On_failure cmd params) ;
 
-        (* Import YUM repository GPG key to check metadata in reposync *)
-        let Pkg_mgr.{cmd; params} = Pkgs.make_cache ~repo_name in
-        ignore (Helpers.call_script cmd params) ;
+        let auth =
```

Comment:
To be clear, it could be `client_auth`.

Reply:
And If the login_with_password bit is handled before constructing `PoolExtHostAuthConf`, it will not be necessary to map `PoolExtHostAuthConf` to `ExtHostAuth`.